workflow-runtime/src/commonMain/kotlin/com/squareup/workflow1/internal/WorkflowRunner.kt

```diff
@@ -72,58 +72,61 @@ internal class WorkflowRunner<PropsT, OutputT, RenderingT>(
   }
 
   /**
-   * Stop processing and go to render on 1 of 3 conditions:
-   * 1. Props have changed.
-   * 2. tick produces an Output to be handled.
-   * 3. The frame 'times out'.
+   * Process the actions queued in the channels of any Workflow in the tree. 'Processing actions'
+   * means handling an event from the UI or from a Worker that will cause state to change.
    *
-   * This function is a gate for when we next render. In the case of a frame timeout or if the
-   * props have been updated, we open the gate and go render, but we have no Output to pass on.
-   * In those cases we return null.
+   * How many actions we process before returning (to go and render()) depends on the
+   * [RuntimeConfig].
+   *
+   * @param waitForAction - whether or not we are suspending until an action is received or only
+   * processing actions that are already readily available. This is only used by the
+   * [RenderingPerFrame] runtime.
    */
   @OptIn(WorkflowExperimentalRuntime::class)
-  suspend fun processActions(): WorkflowOutput<OutputT>? {
-    // First we block and wait until there is an action to process.
-    var processingResult: ActionProcessingResult? = select {
+  suspend fun processActions(waitForAction: Boolean = true): ActionProcessingResult<OutputT>? {
+    // Wait until an action can be applied or the props are updated.
+    var processingResult: ActionProcessingResult<OutputT>? = select {
       onPropsUpdated()
       // Have the workflow tree build the select to wait for an event/output from Worker.
       rootNode.tick(this)
+      if (!waitForAction && runtimeConfig is RenderingPerFrame) {
+        // If we are not waiting for an action to happen, then only wait for [actionWaitMs].
+        onTimeout(runtimeConfig.actionWaitMs) {
+          ActionsExhausted()
+        }
+      }
     }
 
-    if (runtimeConfig is FrameTimeout) {
+    if (runtimeConfig is RenderPassPerFrame) {
       val frameStartTime = currentTimeMillis()
 
       var frameTimeLeft = runtimeConfig.frameTimeoutMs
-      // Then we can process any more actions queued up until a max frame timeout when we should
-      // pass off something to the UI!
-      while (processingResult == null && frameTimeLeft > 0) {
+      // Then we can process any more actions queued up until the frame timeout when we should
+      // pass off something to the UI! If processingResult != null, that means we have an output
+      // for the external system, so we must go back and render then as well. If the result
+      // is ActionsExhausted, there are no actions ready to process.
+      while (
+        processingResult == null &&
+        processingResult !is ActionsExhausted &&
+        frameTimeLeft > 0
```

Comment:
`processingResult == null && processingResult !is ActionsExhausted` is always true.